habitat_baselines/il/trainers/pacman_trainer.py

```diff
@@ -0,0 +1,654 @@
+                            self.device
+                        )
+                        controller_actions_in = controller_actions_in.to(
+                            self.device
+                        )
+                        planner_hidden_idx = planner_hidden_idx.to(self.device)
+                        controller_outs = controller_outs.to(self.device)
+                        controller_action_lengths = (
+                            controller_action_lengths.to(self.device)
+                        )
+                        controller_masks = controller_masks.to(self.device)
+
+                        (
+                            planner_action_lengths,
+                            perm_idx,
+                        ) = planner_action_lengths.sort(0, descending=True)
+                        questions = questions[perm_idx]
+
+                        planner_img_feats = planner_img_feats[perm_idx]
+                        planner_actions_in = planner_actions_in[perm_idx]
+                        planner_actions_out = planner_actions_out[perm_idx]
+                        planner_masks = planner_masks[perm_idx]
+
+                        controller_img_feats = controller_img_feats[perm_idx]
+                        controller_actions_in = controller_actions_in[perm_idx]
+                        controller_outs = controller_outs[perm_idx]
+                        planner_hidden_idx = planner_hidden_idx[perm_idx]
+                        controller_action_lengths = controller_action_lengths[
+                            perm_idx
+                        ]
+                        controller_masks = controller_masks[perm_idx]
+
+                        (
+                            planner_scores,
+                            controller_scores,
+                            planner_hidden,
+                        ) = model(
+                            questions,
+                            planner_img_feats,
+                            planner_actions_in,
+                            planner_action_lengths.cpu().numpy(),
+                            planner_hidden_idx,
+                            controller_img_feats,
+                            controller_actions_in,
+                            controller_action_lengths,
+                        )
+
+                        planner_logprob = F.log_softmax(planner_scores, dim=1)
+                        controller_logprob = F.log_softmax(
+                            controller_scores, dim=1
+                        )
+
+                        planner_loss = planner_loss_fn(
+                            planner_logprob,
+                            planner_actions_out[
+                                :, : planner_action_lengths.max()
+                            ]
+                            .contiguous()
+                            .view(-1, 1),
+                            planner_masks[:, : planner_action_lengths.max()]
```

Comment:
Can this be written cleaner with torch.flatten and expand_dims, I wonder...

Reply:
It's either–
`planner_masks[:,:planner_action_lengths.max()].contiguous().view(-1,1)` (current)
or
`planner_masks[:,:planner_action_lengths.max()].flatten().unsqueeze(1)`

Reply:
You can use `reshape(-1, 1)` if you want a single statement, but it isn't a big deal IMO.